### PR TITLE
added gpt2

### PR DESCRIPTION
**Pull request by kamalojasv181**

### CODE REVIEW

**Comment by kamalojasv181** [comment]
@Ayushk4

File: cformers/interface.py [hunk 1]
```diff
@@ -1,16 +1,13 @@
 """Call's the C++ code from Python."""
 from subprocess import Popen, PIPE
 import hashlib
-import re
 import os
 import sys
-import select
-import wget
```
Comment (Ayushk4) [inline cformers/interface.py:8]: It seems like this has not been updated against the latest push to `master` and is reverting a lot of the changes made
Could you please address this?

File: cformers/interface.py [hunk 2]
```diff
@@ -65,11 +62,14 @@ def get_modes(self):
     'bigscience/bloom-7b1': ModelUrlMap(
         cpp_model_name="bloom",
         int4_fixed_zero="https://huggingface.co/ayushk4/bigscience-.-bloom-7b1/resolve/main/int4_fixed_zero.bin"),
+    'gpt2': ModelUrlMap(
+        cpp_model_name="gpt2",
+        int4_fixed_zero="https://huggingface.co/kamalojasv/gpt2_int4_fixes_zero/resolve/main/model-quant.bin"),
```
Comment (Ayushk4) [inline cformers/interface.py:67]: Can you stick to `int4_fixed_zero.bin` name of the file and `gpt2` as name of the huggingface model repo?

File: cformers/interface.py [hunk 3]
```diff
@@ -108,57 +99,36 @@ def generate(self,
...
-        # Decode the tokens
-        decoded_tokens = self.tokenizer.decode(all_tokens)
+            print(output.strip())
 
         # Wait for the process to finish and return its exit code
-        return {"success": process.wait(),
-                "token_ids": all_tokens,
-                "token_str": decoded_tokens}
-        
+        return process.wait()
```
Comment (Ayushk4) [inline cformers/interface.py:134]: Could you also share cpp files for gpt2 quantization and changes made to main.cpp?

**Review by Ayushk4 — CHANGES_REQUESTED** [review]
Thanks a lot @kamalojasv181 for the PR.
I left a couple of comments. Once you address those, we can test it and merge this PR.

File: cformers/interface.py (hunk 2 above)
Comment (Ayushk4) [reply cformers/interface.py:70]: This should be trivial to edit on the hugginface-interface.

File: cformers/interface.py (hunk 1 above)
Comment (kamalojasv181) [reply cformers/interface.py]: Done

File: cformers/interface.py (hunk 2 above)
Comment (kamalojasv181) [reply cformers/interface.py:70]: Sure

File: cformers/interface.py (hunk 3 above)
Comment (Ayushk4) [reply cformers/interface.py]: Let me know if you need any help regarding it.

Comment (kamalojasv181) [reply cformers/interface.py]: Actually I used the gptj script only

File: cformers/interface.py (hunk 2 above)
Comment (kamalojasv181) [reply cformers/interface.py]: Done

**Comment by kamalojasv181** [comment]
All changes done @Ayushk4

File: cformers/interface.py [hunk 4]
```diff
@@ -65,6 +65,9 @@ def get_modes(self):
     'bigscience/bloom-7b1': ModelUrlMap(
         cpp_model_name="bloom",
         int4_fixed_zero="https://huggingface.co/ayushk4/bigscience-.-bloom-7b1/resolve/main/int4_fixed_zero.bin"),
+    'gpt2': ModelUrlMap(
+        cpp_model_name="gpt2",
+        int4_fixed_zero="https://huggingface.co/kamalojasv/gpt2_int4_fixes_zero/resolve/main/int4_fixed_zero.bin"),
```
Comment (Ayushk4) [inline cformers/interface.py:70]: Can you change it from `kamalojasv/gpt2_int4_fixes_zero` to `kamalojasv/gpt2`?
Also, can you add the CPP quantization script used to quantize this model?

**Review by Ayushk4 — COMMENTED** [review]
Thanks a lot - can you make one minor change to the name and add the gpt2 quantization script as well?

**Comment by kamalojasv181** [comment]
I used the same script as GPT j. I was operating under the assumption that once the model is ported to ggml format, it should be inferable using one unified script. Now that I checked the main.cpp file there is a lot of hard code for each model. I will look into that and see how I can adapt it to gpt2. Thanks

**Comment by kamalojasv181** [comment]
@Ayushk4 is the quantization file correct? It does run and return a model. I just took the gptj quantize file and updated the hyperparameters. Is there anything else required?

**Review by Ayushk4 — COMMENTED** [review]
Thanks. This look good.
After this, we will need to add support for gpt2 in `cformers/cpp/main.cpp`, but we could also have another PR after that.

If so, then you can undo modification to README.

Also, can you change the path of the model from `https://huggingface.co/kamalojasv/gpt2_int4_fixes_zero/resolve/main/int4_fixed_zero.bin` to `https://huggingface.co/kamalojasv/gpt2/resolve/main/int4_fixed_zero.bin` -- model repo name from "gpt2_int4_fixes_zero" -> "gpt2".

**Comment by kamalojasv181** [comment]
Hi @Ayushk4 I am confused, why lmh_g and lmh_b (https://github.com/kamalojasv181/cformers/blob/4b8f2c16ef06b45dafc239f842b0682766f646f9/cformers/cpp/main.cpp#L972) are in the model wiights. I cannot find these weights in the actual model

**Comment by kamalojasv181** [comment]
Also, what is the use of key and value memory? https://github.com/kamalojasv181/cformers/blob/4b8f2c16ef06b45dafc239f842b0682766f646f9/cformers/cpp/main.cpp#L978

**Comment by Ayushk4** [comment]
@kamalojasv181 I have been getting notifications/mail that you commented on this PR, but I am unable to see those comments here.

You could send those again?

**Comment by kamalojasv181** [comment]
Hey @Ayushk4 I think it is done. I also resolved all the merge conflicts. Please give me your thoughts before we can wrap this up. Thanks

**Review by Ayushk4 — APPROVED** [review]

**Pull request merged**